apollo-router/src/protocols/sse/client.rs

```diff
@@ -0,0 +1,640 @@
+    }
+
+    /// Build with an HTTP client connector.
+    pub(crate) fn build_http(self) -> impl Client {
+        self.build_with_conn(HttpConnector::new())
+    }
+
+    /// Build with an HTTPS client connector, using the OS root certificate store.
+    pub(crate) fn build(self) -> impl Client {
+        let conn = hyper_rustls::HttpsConnectorBuilder::new()
+            .with_native_roots()
+            .https_or_http()
+            .enable_http2()
+            .build();
+        self.build_with_conn(conn)
+    }
+
+    /// Build with the given [`hyper::client::Client`].
+    pub(crate) fn build_with_http_client<C>(self, http: hyper::Client<C>) -> impl Client
+    where
+        C: Connect + Clone + Send + Sync + 'static,
+    {
+        ClientImpl {
+            http,
+            request_props: RequestProps {
+                url: self.url,
+                headers: self.headers,
+                method: self.method,
+                body: self.body,
+                reconnect_opts: self.reconnect_opts,
+                max_redirects: self.max_redirects.unwrap_or(DEFAULT_REDIRECT_LIMIT),
+            },
+            last_event_id: self.last_event_id,
+        }
+    }
+}
+
+#[derive(Clone)]
+struct RequestProps {
+    url: Uri,
+    headers: HeaderMap,
+    method: String,
+    body: Option<String>,
+    reconnect_opts: ReconnectOptions,
+    max_redirects: u32,
+}
+
+/// A client implementation that connects to a server using the Server-Sent Events protocol
+/// and consumes the event stream indefinitely.
+/// Can be parameterized with different hyper Connectors, such as HTTP or HTTPS.
+struct ClientImpl<C> {
+    http: hyper::Client<C>,
+    request_props: RequestProps,
+    last_event_id: Option<String>,
+}
+
+impl<C> Client for ClientImpl<C>
+where
+    C: Connect + Clone + Send + Sync + 'static,
+{
+    /// Connect to the server and begin consuming the stream. Produces a
+    /// [`Stream`] of [`Event`](crate::Event)s wrapped in [`Result`].
+    ///
+    /// Do not use the stream after it returned an error!
+    ///
+    /// After the first successful connection, the stream will
+    /// reconnect for retryable errors.
+    fn stream(&self) -> BoxStream<Result<Sse>> {
+        Box::pin(ReconnectingRequest::new(
+            self.http.clone(),
+            self.request_props.clone(),
+            self.last_event_id.clone(),
+        ))
+    }
+}
+
+pin_project! {
+
+    #[project = StateProj]
+    #[allow(clippy::large_enum_variant)] // false positive
+    enum State {
+        New,
+        Connecting {
+            retry: bool,
+            #[pin]
+            resp: ResponseFuture,
+        },
+        Connected {
+            #[pin]
+            body: hyper::Body
+        },
+        WaitingToReconnect {
+            #[pin]
+            sleep: Sleep
+        },
+        FollowingRedirect {
+            redirect: Option<HeaderValue>
+        },
+        StreamClosed,
+    }
+}
+
+impl State {
+    fn name(&self) -> &'static str {
```

Comment:
Should we call this method `as_str()` to follow the Rust convention ? Or is it more accurate to keep `name` ? Just asking I don't have strong opinion

Reply:
I changed it to `as_str`